### PR TITLE
Implement TestController. Implement t.click() (closes #415)

### LINKED ISSUE #415: Implement TestController
- [x] Refactor command
- [x] Move command options tests to server
- [x] Implement basic interface
- [x] Pass it to runner
- [x] Implement action execution
- [x] Implement options validation (#420)
- [x] Implement chaining
- [x] Test `t.click()`
  - [x] Execution
  - [x] Options validation
  - [x] Selector validation
- [x] Test controller
  - [x] Chaining
  - [x] Callsites correctness in chained call
  - [x] Await tracking
- [x] Implement `await` tracking

### PR DESCRIPTION
**Pull request by inikulin**
\cc @AlexanderMoskovkin @helen-dikareva 
\cc @VasilyStrelyaev  please check strings

### CODE REVIEW

File: src/api/globals/test.js [hunk 1]
```diff
@@ -14,7 +15,21 @@ export default class Test {
...
 
         this.name    = name;
-        this.fn      = fn;
         this.fixture = fixture;
+        this.fn      = Test._createTestFunction(fn);
+    }
+
+    static _createTestFunction (fn) {
+        return async testRun => {
+            // NOTE: fn() result used for the testing purposes
```
Comment (VasilyStrelyaev) [inline src/api/globals/test.js]:
```
            // NOTE: fn() result used for testing purposes
```

File: src/api/globals/test.js [hunk 2]
```diff
@@ -14,7 +15,21 @@ export default class Test {
...
+            var controller = new TestController(testRun);
+            var result     = await fn(controller);
+
+            // NOTE: check if last command in the test
+            // function missing `await` keyword.
```
Comment (VasilyStrelyaev) [inline src/api/globals/test.js]:
```
            // NOTE: check if the last command in the test
            // function is missing the `await` keyword.
```

File: src/api/test-controller.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    _createAPIMethodsList () {
+        return Object
+            .keys(TestController.prototype)
+            .filter(name => !/^_/.test(name) && typeof TestController.prototype[name] === 'function');
+    }
+
+
+    // NOTE: we track missing `awaits` by exposing special custom Promise to user code.
```
Comment (VasilyStrelyaev) [inline src/api/test-controller.js]:
```
    // NOTE: we track missing `awaits` by exposing a special custom Promise to user code.
```

File: src/api/test-controller.js [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    // Action or assertion is awaited if:
+    // a)someone used `await` so Promise's `then` function executed
+    // b)Promise chained by using one of the mixed-in controller methods
+    //
+    // In both scenarios we check that callsite that produced Promise is equal to the one
+    // which currently missing await. This is required to workaround scenarios like this:
```
Comment (VasilyStrelyaev) [inline src/api/test-controller.js]:
```
    // In both scenarios, we check that callsite that produced Promise is equal to the one
    // that is currently missing await. This is required to workaround scenarios like this:
```

File: test/functional/fixtures/api/es-next/test-controller/test.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
+var expect = require('chai').expect;
+
+// NOTE: we run tests in chrome only, because we mainly test server API functionality.
+// Actions functionality is tested in lower-level raw API.
+describe('[API] TestController', function () {
+    it('Should should support chaining [ONLY:chrome]', function () {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/api/es-next/test-controller/test.js]:
```
    it('Should support chaining [ONLY:chrome]', function () {
```

File: test/functional/fixtures/api/es-next/test-controller/test.js [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        return runTests('./testcafe-fixtures/test-controller-test.js', 'Chaining', { shouldFail: true })
+            .catch(function (err) {
+                expect(err).to.contains('-btn1-btn2-btn3-page2-btn1-page2-btn2');
+            });
+    });
+
+    it('Should should produce correct callsites for chained calls [ONLY:chrome]', function () {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/api/es-next/test-controller/test.js]:
```
    it('Should produce correct callsites for chained calls [ONLY:chrome]', function () {
```

File: test/functional/fixtures/api/raw/click/test.js [hunk 7]
```diff
@@ -53,14 +55,32 @@ describe('Run click from a raw file', function () {
...
+
+    it('Should fail if action has incorrect selector', function () {
+        return runTests('./testcafe-fixtures/click.testcafe', 'Incorrect action selector', { shouldFail: true })
+            .catch(function (err) {
+                expect(err).eql('Action selector is expected to be a string, but it was number.  ' +
+                                '[[Incorrect action selector callsite]]');
+            });
+    });
+
+    it('Should fail if action has option', function () {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/api/raw/click/test.js]: *incorrect* option you mean?
```
    it('Should fail if action has an incorrect option', function () {
```

File: test/functional/fixtures/api/es-next/click/testcafe-fixtures/click-test.js [hunk 8]
```diff
@@ -0,0 +1,16 @@
+// NOTE: to preserve callsites add new tests AFTER existing
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/api/es-next/click/testcafe-fixtures/click-test.js]:
```
// NOTE: to preserve callsites, add new tests AFTER the existing ones
```

File: test/functional/fixtures/api/es-next/test-controller/testcafe-fixtures/test-controller-test.js [hunk 9]
```diff
@@ -0,0 +1,65 @@
+import { missingAwaitFn } from './helpers';
+// NOTE: to preserve callsites add new tests AFTER existing
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/api/es-next/test-controller/testcafe-fixtures/test-controller-test.js]:
```
// NOTE: to preserve callsites, add new tests AFTER the existing ones
```

**Comment by VasilyStrelyaev** [comment]
\r-

**Comment by inikulin** [comment]
FPR

**Comment by VasilyStrelyaev** [comment]
lgtm

File: src/api/test-controller.js [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    // b)Promise chained by using one of the mixed-in controller methods
+    //
+    // In both scenarios, we check that callsite that produced Promise is equal to the one
+    // that is currently missing await. This is required to workaround scenarios like this:
+    //
+    // var t2 = t.click('#btn1'); // <-- stores new callsiteWithoutAwait
+    // await t2;                  // <-- callsiteWithoutAwait = null
+    // t.click('#btn2');          // <-- stores new callsiteWithoutAwait
+    // await t2.click('#btn3');   // <-- without check it will set callsiteWithoutAwait = null, so we will lost tracking
+    _createExtendedPromises (promise, callsite) {
```
Comment (inikulin) [inline src/api/test-controller.js]: Should be renamed to `_createExtendedPromise`

File: src/errors/test-run/type.js [hunk 11]
```diff
@@ -7,11 +7,12 @@ export default {
     actionNumberOptionError:         'actionNumberOptionError',
     actionPositiveNumberOptionError: 'actionPositiveNumberOptionError',
     actionBooleanOptionError:        'actionBooleanOptionError',
-    actionSelectorTypeError:         'actionSelectorTypeError',
     actionOptionsTypeError:          'actionOptionsTypeError',
 
     uncaughtErrorOnPage: 'uncaughtErrorOnPage',
+    missingAwaitError:   'missingAwaitError',
```
Comment (helen-dikareva) [inline src/errors/test-run/type.js]: Wrong formatting

File: test/functional/fixtures/api/es-next/test-controller/testcafe-fixtures/helpers.js [hunk 12]
```diff
@@ -0,0 +1,3 @@
+export function missingAwaitFn (t) {
```
Comment (helen-dikareva) [inline test/functional/fixtures/api/es-next/test-controller/testcafe-fixtures/helpers.js]: We really need export this function to another file? Now it used only ones, will we use it again later?

File: src/api/test-controller.js [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+
+        extendedPromise.then = function () {
+            ensureAwait();
+            return originalThen.apply(this, arguments);
+        };
+
+        this.apiMethods.forEach(name => {
+            var controller = this;
+
+            extendedPromise[name] = function () {
```
Comment (helen-dikareva) [inline src/api/test-controller.js]: May be we can use arrow function here, but it's not so important

File: test/functional/fixtures/api/es-next/test-controller/testcafe-fixtures/helpers.js (hunk 12 above)
Comment (inikulin) [inline test/functional/fixtures/api/es-next/test-controller/testcafe-fixtures/helpers.js]: It's a test

Comment (helen-dikareva) [inline test/functional/fixtures/api/es-next/test-controller/testcafe-fixtures/helpers.js]: I see

**Comment by AlexanderMoskovkin** [comment]
lgtm

File: src/api/test-controller.js (hunk 13 above)
Comment (inikulin) [inline src/api/test-controller.js]: We use `arguments` in both cases. Arrows doesn't have `arguments` object.

**Comment by inikulin** [comment]
FPR

File: src/api/test-controller.js (hunk 13 above)
Comment (helen-dikareva) [inline src/api/test-controller.js]: ok

**Comment by helen-dikareva** [comment]
lgtm

**Pull request merged**